### ISSUE TITLE
mem_top() gets stuck

### ISSUE THREAD

**Issue by greatvovan** [denis-ryzhkov/mem_top#8]
I am running a FastAPI service accessing MS SQL DB through sqlalchemy. I injected mem_top() in the middleware like this:

```python
@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
    response = await call_next(request)
    logger.debug('Calling mem_top()...')
    logging.debug(mem_top())
    return response
```
But the call to `mem_top()` gets stuck either working for too long (I waited ~15 minutes) or falling into an infinite loop, having printed nothing after "Calling mem_top()...". The amount of memory consumed by the process at the time of the call is around 80 MB.

I could not reproduce the behaviour using a trivial FastAPI app. Not sure how worthy this bug report, I understand that the behavior might be dependent on the memory structures in use or the amount of them, but may be there known issues in the lib or you have an idea of what's going on?

<details>
  <summary>Environment</summary>

MacOS 13.6.3
Python 3.12
fastapi==0.99.1
uvicorn==0.23.0
SQLAlchemy==2.0.21
starlette==0.27.0
</details>

**Comment by denis-ryzhkov** [comment]
Could you please copy the source code of mem_top into your service, adding more logs inside of it?

**Comment by greatvovan** [comment]
Oh, right, I did not realize it is, in fact, a one-pager script.

So it gets stuck here, on yielding this iterable: https://github.com/denis-ryzhkov/mem_top/blob/master/mem_top.py#L79

I expanded it into an explicit loop like this:

```python
    refs = []
    for i, obj in enumerate(objs):
        print('printing object')
        print(i, type(obj), _repr(obj)[:100])
        print('get_referents')
        referents = gc.get_referents(obj)
        print('len')
        length = len(referents)
        print('tuple')
        tup = (length, obj)
        print('append')
        refs.append(tup)
```

What I am observing is quite crazy to say the least.

The loop quite quickly passes through around 92k objects and eventually gets stuck like this:

```
printing object
92451 <class 'tuple'> (<class 'webbrowser.UnixBrowser'>,)
get_referents
len
tuple
```

- The iteration count is always around ~92450 (+/- 10). The total number of objects is ~118k.
- The last row printed is always "tuple".

It looks like at some point the interpreter is unable to create a tuple. The process get stuck forever in such a state and the CPU is at ~100% at one core.

I am about to think that it has nothing to do with your library and it's a Python bug or a corrupted installation.

**Comment by greatvovan** [comment]
I installed another version of Python through pyenv and it is working now. Closing...

**Comment by denis-ryzhkov** [comment]
Thanks! If you or someone else faces this issue again, please add `flush=True` to the print statements above as Python may get stuck before flushing stdout. So it was probably not a tuple issue.

https://docs.python.org/3/library/functions.html#print

**Comment by greatvovan** [comment]
I ran the test with [PYTHONUNBUFFERED=1](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONUNBUFFERED), which should enforce such a behaviour. Plus, it got stuck on this line every time despite slightly different output, whereas a buffer cutting point expected to be pretty random. I tested it in different ways, this snippet was one of them, every time it would get stuck on `(length, obj)`.

I know it sounds pretty insane, I am surprised not less than you. I am prone to blame a corrupted build/installation, I already encountered crazy things in the past that were fixed after running on a different interpreter (even the same version).
